### PR TITLE
Fix flaky test and avoid divide by zero in SoftmaxCrossEntropyLoss-CPU.

### PR DESCRIPTION
**Pull request by codemzs**
The change fixes two issues that were caught as _flaky_ test:

1. **Invalid label data being passed into tests**
    - `random.Uniform` returns values that are outside the range [a, b) because of `float` to `int64` cast internally due to the usage of `uniform_real_distribution` that expects template type of floating types only. This results in index out of bounds exception when we use `uniform_real_distribution` with `std::default_random_engine` and perform an explicit cast from `float` to `int64`. We must have an overload of  `random.Uniform` that uses `uniform_int_distribution` to prevent the need for such a cast.

2. **SoftmaxCrossEntropyLoss-CPU dividing by zero**
    - One of the test was exercising the scenario where all label values were `0` and `ignore_index` was also `0` hence the loss should also be zero since all values should be dropped. On GPU the kernel correctly calculated loss as zero but on CPU it ended up dividing sum of all losses (zero) by zero which resulted in -nan.

Interestingly, both the tests ran fine on Windows even when I ran them in a loop for 100 times. In the hindsight, it seems this issue also existed in _GatherGradOpTest.Gather_axis1_float_impl2_ and _GatherGradOpTest.Gather_axis0_float_impl2_ that I have fixed as part of this PR.

### CODE REVIEW

File: orttraining/orttraining/test/training_ops/cuda/cross_entropy_test.cc [hunk 1]
```diff
@@ -313,7 +313,7 @@ static void TestSoftmaxCrossEntropyLoss(const std::vector<int64_t>* X_dims,
 TEST(CudaKernelTest, SoftmaxCrossEntropyLoss_TinySizeTensor) {
   std::vector<int64_t> X_dims{8, 2};
   std::vector<int64_t> index_dims{8};
-  std::vector<int64_t> weight_dims{8};
+  std::vector<int64_t> weight_dims{2};
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/training_ops/cuda/cross_entropy_test.cc:316]: changed by accident?

File: onnxruntime/test/common/tensor_op_test_utils.h [hunk 2]
```diff
@@ -29,6 +29,17 @@ class RandomValueGenerator {
     return val;
   }
 
+  template <class T>
```
Comment (SherlockNoMad) [inline onnxruntime/test/common/tensor_op_test_utils.h:32]: a comment explaining the function above is for float [a, b) and this function is for int [a, b]

File: orttraining/orttraining/training_ops/cpu/loss/softmax_cross_entropy_loss.cc [hunk 3]
```diff
@@ -194,7 +193,7 @@ Status SoftmaxCrossEntropyLoss<T1, T2>::Compute(OpKernelContext* context) const
     if (reduction_ != ReductionType::NONE) {
       // Sum loss over n_d samples
       math::Sum<T1, CPUMathUtil>(n_d, loss_sample, loss_data, nullptr);
-      if (reduction_ == ReductionType::MEAN) {
+      if ((reduction_ == ReductionType::MEAN) && (unignored_samples != 0)) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/loss/softmax_cross_entropy_loss.cc:196]: In the case there is no valid sample in this batch, there is no way to compute the gradients. 
We should error out in the case.

Comment (SherlockNoMad) [reply orttraining/orttraining/training_ops/cpu/loss/softmax_cross_entropy_loss.cc:196]: Also, do we have this problem for CUDA kernels?

File: onnxruntime/test/common/tensor_op_test_utils.h (hunk 2 above)
Comment (edgchen1) [reply onnxruntime/test/common/tensor_op_test_utils.h:32]: perhaps we can use the signature: `vector<T> Uniform(const vector<int64_t>& dims, T min, T max)`
and dispatch to integral or floating point implementations with some template metaprogramming (e.g., with std::is_integral and std::is_floating_point).
if we keep them separate, naming them differently may be helpful. for example, UniformReal and UniformInt

**Comment by yuslepukhin** [comment]
Excellent work!

File: orttraining/orttraining/test/training_ops/cuda/cross_entropy_test.cc (hunk 1 above)
Comment (codemzs) [reply orttraining/orttraining/test/training_ops/cuda/cross_entropy_test.cc:316]: No, weights, should be 2, it should be same shape as C

File: orttraining/orttraining/training_ops/cpu/loss/softmax_cross_entropy_loss.cc (hunk 3 above)
Comment (codemzs) [reply orttraining/orttraining/training_ops/cpu/loss/softmax_cross_entropy_loss.cc:196]: We don't have issue on CUDA because there we don't really divide total loss by zero as we have normalize_factor which will be zero but it won't be used at all as all values will be dropped (since target value will be same as ignore_index) and in the end we will do a reduce sum on all zero loss values which will result in a zero. The UT verifies this.  

Why is a gradient of zero not valid when there are no valid values?

File: onnxruntime/test/common/tensor_op_test_utils.h (hunk 2 above)
Comment (codemzs) [reply onnxruntime/test/common/tensor_op_test_utils.h:32]: @edgchen1 I'm not sure that is a clean design because for floating types we will need to use `uniform_real_distribution` but for integer types we will need to use `uniform_int_distribution`. What do you think?

Comment (edgchen1) [reply onnxruntime/test/common/tensor_op_test_utils.h:32]: yes, i was suggesting to keep a consistent signature, i.e., take T min and max values and return vector<T>, but use `uniform_real_distribution` or `uniform_int_distribution` in the implementation depending on T.

Comment (edgchen1) [reply onnxruntime/test/common/tensor_op_test_utils.h:32]: i'm also fine with renaming this function to `UniformInt` or similar, or adding a comment as Sherlock suggested. otherwise it may be hard to easily see the difference between the two Uniform()'s

File: orttraining/orttraining/test/training_ops/cpu/tensor/gather_grad_op_test.cc [hunk 4]
```diff
@@ -105,7 +105,7 @@ TEST(GatherGradOpTest, Gather_axis1_float_impl2) {
   int64_t output_shape = 4;
   RandomValueGenerator random{};
   std::vector<float> grad(random.Uniform<float>({axis_0 * axis_1 * axis_2}, 1.0f, 1.0f));
-  std::vector<int64_t> indices(random.Uniform<int64_t>({axis_1 * axis_2}, (int64_t)0, (int64_t)3));
+  std::vector<int64_t> indices(random.Uniform<int64_t>({axis_1 * axis_2}, (int64_t)0, (int64_t)2));
```
Comment (edgchen1) [inline orttraining/orttraining/test/training_ops/cpu/tensor/gather_grad_op_test.cc:108]: nit: prefer to avoid C-style casts (https://google.github.io/styleguide/cppguide.html#Casting)

Comment (codemzs) [reply orttraining/orttraining/test/training_ops/cpu/tensor/gather_grad_op_test.cc:108]: Good idea! Love static cast syntax in addition to other benefits :)

File: orttraining/orttraining/training_ops/cpu/loss/softmax_cross_entropy_loss.cc (hunk 3 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cpu/loss/softmax_cross_entropy_loss.cc:196]: If all labels are dropped because of `ingore_index`, the loss and gradient should be 0. Let's consider a math equation of `ignore_index=0` and all labels are 0. The loss is actually
```
0 * loss(label[0], predict[0]) + 0 * loss(label[1], predict[1]) + ... + 0 * loss(label[L-1], predict[L-1])
```
where 
```
label[0] = label[1] = ... = label[L-1]
```
Because of the `0*` before individual loss term, their total loss is 0 and their gradient will be zero 0 (taking derivative against a constant 0 w.r.t. any tensor would be 0).

A batch with all label being `ignored` is still a valid data point. It can appear in user's data set, so we can't throw.

Comment (codemzs) [reply orttraining/orttraining/training_ops/cpu/loss/softmax_cross_entropy_loss.cc:196]: Thanks, @wschin !!!

**Comment by codemzs** [comment]
Thanks @yuslepukhin ! I have resolved all the comments and have a sign-off, once builds complete we can merge. Thank you for your patience.

File: onnxruntime/test/common/tensor_op_test_utils.h (hunk 2 above)
Comment (codemzs) [reply onnxruntime/test/common/tensor_op_test_utils.h:34]: Yea that seems a little messy to me. I had rather have it overloaded.

---
In reply to: [416953659](https://github.com/microsoft/onnxruntime/pull/3734#discussion_r416953659) [](ancestors = 416953659)

Comment (codemzs) [reply onnxruntime/test/common/tensor_op_test_utils.h:34]: I have added the comment.

---
In reply to: [416956029](https://github.com/microsoft/onnxruntime/pull/3734#discussion_r416956029) [](ancestors = 416956029)
